keyboards/albert_smart/albert_smart84/albert_smart84.c

```diff
@@ -0,0 +1,339 @@
+    for(uint8_t x=0; x<MATRIX_ROWS; x++){
+        pin_t pin;
+        pin = row_pins[x];
+        if (pin != NO_PIN){
+            setPinInputHigh(pin);
+            palEnableLineEvent(pin, PAL_EVENT_MODE_BOTH_EDGES);
+        }
+    }
+#else
+
+    const long unsigned int row_pins[] = MATRIX_ROW_PINS;
+    for(uint8_t x=0; x<MATRIX_ROWS; x++){
+        pin_t pin;
+        pin = row_pins[x];
+        if (pin != NO_PIN){
+            setPinOutput(pin);
+            writePinLow(pin);
+        }
+    }
+
+    const long unsigned int col_pins[] = MATRIX_COL_PINS;
+    for(uint8_t x=0; x<MATRIX_COLS; x++){
+        pin_t pin;
+        pin = col_pins[x];
+        setPinInputHigh(pin);
+        palEnableLineEvent(pin, PAL_EVENT_MODE_BOTH_EDGES);
+    }
+#endif
+    /* going to anabiosis*/
+    //chSysLock();
+    /* Clear Wake-up flag */
+    palEnableLineEvent(A6, PAL_EVENT_MODE_BOTH_EDGES);
+    palEnableLineEvent(B14, PAL_EVENT_MODE_BOTH_EDGES);
+    _pal_lld_enablepadevent(0,18,PAL_EVENT_MODE_BOTH_EDGES);
+    PWR->CR |= PWR_CR_CWUF | PWR_CR_CSBF;
+
+    /* Select Sleep mode */
+    //PWR->CR |= PWR_CR_PDDS | PWR_CR_LPDS; //  Can NOT wake up
+    /* Set SLEEPDEEP bit of Cortex System Control Register */
+    SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;
+    
+    /* Request Wait For Interrupt */
+    __WFI();
+    EXTI->IMR &= ~(1<<18);
+    EXTI->EMR &= ~(1<<18);
+    EXTI->RTSR &= ~(1<<18);
+    EXTI->FTSR &= ~(1<<18);
+    EXTI->PR    = 1<<18;
+    /* Reset SLEEPDEEP bit of Cortex System Control Register */
+    SCB->SCR &= ~SCB_SCR_SLEEPDEEP_Msk;
+
+    stm32_clock_init();
+    matrix_init();
+    usbWakeupHost(&USB_DRIVER);  //解决休眠无法唤醒的问题
+    restart_usb_driver(&USB_DRIVER);
+
+    setPinOutput(LED1);
+    setPinOutput(LED2);
+    setPinOutput(LED_EN);
+
+    loop10hz_token = defer_exec(LOOP_10HZ_PERIOD, loop_10Hz, NULL);
+}
+
+uint8_t sleep_flag = 0;
+
+void sleep_mode(void){
+
+    if(usb_suspend_flag == true){
+        if(timer_elapsed32(usb_suspend_time) > 500){
+            POWER_EnterSleep();
+            usb_suspend_flag = false;
+        }
+    }
+    if(USB_DRIVER.state == USB_SUSPENDED){
+        if(sleep_flag == 1){
+            usb_suspend_time = timer_read32();
+            sleep_flag = 0;
+            usb_suspend_flag = true;
+        }
+    }
+    if(USB_DRIVER.state == USB_ACTIVE){
+        // usb_suspend_flag = false;
+        sleep_flag = 1;
+    }
+    else{
+        sleep_flag = 2;
+    }
+}
+
+void keyboard_pre_init_kb(void) {
+    setPinInputLow(A9);
+    setPinInputLow(A10);
+    setPinInput(LED_EN);
+
+    setPinInputHigh(A7);
+    setPinInputHigh(A6);
+}
+
+void keyboard_post_init_kb(void) {
+
+
+    AFIO->MAPR = (AFIO->MAPR & ~AFIO_MAPR_SWJ_CFG_Msk);
+    AFIO->MAPR|= AFIO_MAPR_SWJ_CFG_DISABLE;
+
+    setPinOutput(LED1);
+    setPinOutput(LED2);
+    setPinOutput(LED_EN);
+    writePinLow(LED_EN);
+    writePinLow(LED1);
+    writePinLow(LED2);
+
+    setPinInputHigh(B14);
+
+    lock_win = dynamic_keymap_get_keycode(0,6,3);
```

Comment:
Is there a reason for this, aside from duplicating the magic keys' no-gui functionality?

Reply:
In addition to this, it is also use to power-down save which keyboard layer in use.

Reply:
In theory, the default layer state should do that, especially if you're saving it to persistent storage (eeprom). 

Additionally, tthe dynamic_keymap function shouldn't be called, period.  `keycode_at_keymap_location` should be used here.  If the dynamic keymap feature is enabled, then it uses that, otherwise, it uses the stock method.   And since you can't be sure that it's always enabled (and don't perform any checks).